lib/ansible/module_utils/facts/hardware/linux.py

```diff
@@ -443,51 +445,81 @@ def _mtab_entries(self):
             mtab_entries.append(fields)
         return mtab_entries
 
-    @timeout.timeout()
+    def get_mount_info(self, mount_info, uuids):
+
+            mount_info.update(get_mount_size(mount_info['mount']))
+
+            # _udevadm_uuid is a fallback for versions of lsblk <= 2.23 that don't have --paths
+            # see _run_lsblk() above
+            # https://github.com/ansible/ansible/issues/36077
+            mount_info['uuid'] = uuids.get(mount_info['device'], self._udevadm_uuid(mount_info['device']))
+
+            return mount_info
+
     def get_mount_facts(self):
-        mount_facts = {}
 
-        mount_facts['mounts'] = []
+        mounts = []
 
+        # gather system lists
         bind_mounts = self._find_bind_mounts()
         uuids = self._lsblk_uuid()
         mtab_entries = self._mtab_entries()
 
-        mounts = []
+        # start threads to query each mount
+        results = {}
+        pool = ThreadPool(processes=min(len(mtab_entries), cpu_count()))
+        maxtime = globals().get('GATHER_TIMEOUT') or timeout.DEFAULT_GATHER_TIMEOUT
         for fields in mtab_entries:
-            device, mount, fstype, options = fields[0], fields[1], fields[2], fields[3]
 
-            if not device.startswith('/') and ':/' not in device:
-                continue
+            device, mount, fstype, options = fields[0], fields[1], fields[2], fields[3]
 
-            if fstype == 'none':
+            if not device.startswith('/') and ':/' not in device or fstype == 'none':
                 continue
 
-            mount_statvfs_info = get_mount_size(mount)
-
-            if mount in bind_mounts:
-                # only add if not already there, we might have a plain /etc/mtab
-                if not self.MTAB_BIND_MOUNT_RE.match(options):
-                    options += ",bind"
-
-            # _udevadm_uuid is a fallback for versions of lsblk <= 2.23 that don't have --paths
-            # see _run_lsblk() above
-            # https://github.com/ansible/ansible/issues/36077
-            uuid = uuids.get(device, self._udevadm_uuid(device))
-
             mount_info = {'mount': mount,
                           'device': device,
                           'fstype': fstype,
-                          'options': options,
-                          'uuid': uuid}
+                          'options': options}
 
-            mount_info.update(mount_statvfs_info)
-
-            mounts.append(mount_info)
+            if mount in bind_mounts:
+                # only add if not already there, we might have a plain /etc/mtab
+                if not self.MTAB_BIND_MOUNT_RE.match(options):
+                    mount_info['options'] += ",bind"
+
+            results[mount] = {'info': mount_info,
+                              'extra': pool.apply_async(self.get_mount_info, (mount_info, uuids)),
+                              'start': time.time() + maxtime}
+
```

Comment:
You don't need start with this code as it will be the same for all the code.  You can just use a single scalar local variable to hold the value.  (It also is the end time or max time (and maxtime is really max_timeout), not start).

Adding that together with the note that we should probably process statvfs information separate from the uuid info:
``` python
results[mount] = {'info': mount_info,
                            'statvfs': pool.apply_async(get_mount_size, (mount_info['mount'],)),
                            'uuid': uuids.get(mount_info['device']) or pool.apply_async(self._udevadm_uuid, (mount_info['device'],)),}
max_time = time.time() + max_timeout
```

Reply:
not as single scalar since we are tracking timeout PER child , otherwise we could just leave the general timeout at top of the method

Reply:
You are not tracking timeout per child with this code.  The  time you're recording here is when you ask the library to queue the work.... It's not when the work starts.  Therefore, you might as well save a single general timeout at the top of the method.

Reply:
well, yes, queueing per child to be exact, in some of my most extreme tests this was up to 2secs diff, so i opted to keep it per child

Reply:
I was looking at time.clock() at the begining of the threaded function as that would give us a more accurate 'execution time'  but we cannot compare it before the `.get` which is the thing we skip on timeout.

Reply:
What was your extreme test?

Reply:
hosts with +300 mount points

Reply:
I think you should rerun your test... The cost to invoke apply_async() on 300 elements is .01 seconds here. Adding in the regex, it's still only .05 seconds.  .05 seconds on a 10 second timeout is virtually nothing.

Reply:
Close to what i see when its a single target, but i'm testing on 50-70 vms on same host (2cpu) with variable mount points (shell below creates them) in parallel, which is not a fair test. 

I was trying to find a point in which i could check the difference in performance between threaded/non threaded and I ended up with this, which also showed me the queuing diff can get big on busy system. The check themselves need more variety, seems like bind mounts end up caching the info making it a non optimal test, still trying to figure out something i can whip up locally.

```bash
#!/bin/bash

ADD=$((${RANDOM}%100))
LIM=$((250+${ADD}))

for i in $(seq 1 $LIM)
do
        mkdir /mnt/test_$i
        mount -o bind /var /mnt/test_$i
done
```